### PR TITLE
Feature signup

### PR DESCRIPTION
**Pull request by jonasKjellerup**
The pull request addresses issue #254 

The pull request adds the following:

- A signup form (linked to via "Tilmeld" button on the index page)
- Adds a `Command` for automatically processing signup payments

The new signup flow would be as follows:

1. The user fills out the signup form.
2. On success the user will be redirected to a status page with a clickable (for mobile users) QR-code.
3. Once payment has been registered the account will be activated and a welcome mail will be sent.

### CODE REVIEW

File: stregsystem/views.py [hunk 1]
```diff
@@ -465,3 +472,43 @@ def qr_payment(request):
     data = 'mobilepay://send?{}'.format(urllib.parse.urlencode(query))
 
     return qr_code(data)
+
+
+def signup(request):
+    is_post = request.method == "POST"
+    form = SignupForm(request.POST) if is_post else SignupForm()
```
Comment (falkecarlsen) [inline stregsystem/views.py:479]: Redundant, can be reduced to one statement. If request.POST, then it is post.

File: stregsystem/forms.py [hunk 2]
```diff
@@ -35,6 +35,19 @@ class PurchaseForm(forms.Form):
     product_id = forms.IntegerField()
 
 
+class SignupForm(forms.Form):
+    username = forms.CharField(max_length=16, label='Brugernavn',
+                               error_messages={'required': 'Udfyldning af `Brugernavn` er påkrævet.',
+                                               'max_length': 'Længden af `Brugernavn må ikke overstige 16 tegn.`'})
```
Comment (Nielswps) [inline stregsystem/forms.py:41]: Misplacement of '\`', should be 'Længden af `Brugernavn` må ikke overstige 16 tegn.'

File: stregsystem/forms.py [hunk 3]
```diff
@@ -35,6 +35,19 @@ class PurchaseForm(forms.Form):
...
+    email = forms.EmailField(label='E-Mail')
+    firstname = forms.CharField(max_length=20, label='Fornavn',
+                                error_messages={'required': 'Udfyldning af `Fornavn` er påkrævet.',
+                                                'max_length': 'Længden af `Fornavn må ikke overstige 16 tegn.`'})
```
Comment (Nielswps) [inline stregsystem/forms.py:45]: Same as above, should be 'Længden af `Fornavn` må ikke overstige 16 tegn.'

File: stregsystem/forms.py [hunk 4]
```diff
@@ -35,6 +35,19 @@ class PurchaseForm(forms.Form):
...
+    lastname = forms.CharField(max_length=30, label='Efternavn',
+                               error_messages={'required': 'Udfyldning af `Efternavn` er påkrævet.',
+                                               'max_length': 'Længden af `Efternavn må ikke overstige 16 tegn.`'})
```
Comment (Nielswps) [inline stregsystem/forms.py:48]: Same as above, should be 'Længden af `Efternavn` må ikke overstige 16 tegn.'

File: stregsystem/migrations/0016_auto_20220131_2109.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        migrations.AddField(
+            model_name='member',
+            name='signup_due_paid',
+            field=models.BooleanField(default=True),
+        ),
+        migrations.CreateModel(
+            name='PendingSignup',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('due', models.IntegerField(default=20000)),
```
Comment (Nielswps) [inline stregsystem/migrations/0016_auto_20220131_2109.py:24]: Consider adding a comment about what 20000 represents. I assume that it is 200,00 DKK in øre?

File: stregsystem/models.py [hunk 6]
```diff
@@ -661,3 +666,30 @@ def __unicode__(self):
 
     def __str__(self):
         return self.title + " -- " + str(self.pub_date)
+
+
+class PendingSignup(models.Model):
+    member = models.ForeignKey(Member, on_delete=models.CASCADE, null=False)
+    due = models.IntegerField(default=200 * 100)
```
Comment (Nielswps) [inline stregsystem/models.py:673]: IMO: Either write 20000 as in `stregsystemet/models.py` or do this there as well, to be consistent

**Review by Nielswps — CHANGES_REQUESTED** [review]
I think this would be a brilliant feature during events for members where non-members might attend, wanting to signup, good job :+1: 
Just some minor string typos

File: stregsystem/models.py (hunk 6 above)
Comment (jonasKjellerup) [reply stregsystem/models.py:685]: I assume you mean as in `[stregsystem/migrations/0016_auto_20220131_2109.py](https://github.com/f-klubben/stregsystemet/pull/297/files/90af2f04a8eb117a6597404182d6616b5036601b#diff-ff4bbb2f164add9b3111918d75b6a5830763528bfe98589659b83c3c58099373)`. The migrations files are auto generated, hence the discrepancy based on the class from the model file, so it doesn't really make sense to edit them.

**Comment by krestenlaust** [comment]
Fixes #254

File: stregsystem/models.py [hunk 7]
```diff
@@ -673,3 +678,30 @@ def __unicode__(self):
...
```
Comment (krestenlaust) [inline stregsystem/models.py:685]: Wouldn't it be better to define the membership cost outside of the model to not require migrations in-case of vedtægtsændringer?

**Comment by krestenlaust** [comment]
I've just looked into this implementation to make sure that #407 doesn't make this outdated, it doesn't look like it. It just needs to catch up with upstream (66 commits behind)

**Comment by JakobTopholt** [comment]
This pull request would be a godsent.

File: stregsystem/models.py (hunk 7 above)
Comment (krestenlaust) [reply stregsystem/models.py:682]: I guess it's just a default

**Comment by krestenlaust** [comment]
Working towards the following feature:
At a meeting with the forstyrrelse, it was decided that it would be preferable to have a view like MobilePayTool to accept signups manually

**Review by krestenlaust — COMMENTED** [review]
Initial reviews are welcome🙌

Major changes in this PR:
 - Introduced new model PendingSignup, which only exists to track members who isn't approved and hasn't paid membership due.
 - Turned existing Mobilepaytool abstract, and implemented a signup tool with the same concept.
 - Some refactoring of the AutoPayment-command to properly process paying towards membership dues.

Missing features:
 - [ ] 'Area of Study'-choice in the signup form to populate 'notes' of the member.
 - [ ] ~~'Reject'-functionality for Signup-Tool hasn't been implemented yet.~~ **Feature rejected for now.**
 - [ ] ~~Haven't added QR-code to error_signupdue-screen (Presented when trying to access user menu with member without due paid)~~ **This feature won't be implemented here.**

Known bugs:
 - [x] ~~Approving a member who has already paid their due, will not remove the pending signup-object.~~ **Fixed**
 - [ ] I also suspect there are some issues related to email triggers, depending on the order a user is approved and due is paid.

I'll try to add some more email tests

**Comment by krestenlaust** [comment]
Feature is finished! I've decided to ignore the potential bug with welcome emails for now.

**Comment by BoAnd** [comment]
As I see it the sign up logic works. There is some small problem with the differences sign up templates.
Like we could merge template **error_signupdue.html** and **error_signup_not_approved.html** into one template. 

But I think it could be nice to just merge this into prod and the fix the problems afterwards (If there is any :dagger:).

**Review by BoAnd — APPROVED** [review]

**Comment by krestenlaust** [comment]
> As I see it the sign up logic works. There is some small problem with the differences sign up templates. Like we could merge template **error_signupdue.html** and **error_signup_not_approved.html** into one template.
> 
> But I think it could be nice to just merge this into prod and the fix the problems afterwards (If there is any 🗡️).

Thank you for your review! I've made another issue with your note

**Pull request merged**